exec/cnex/number.c

```diff
@@ -17,49 +17,88 @@
 char *number_to_string(Number x)
 {
     static char buf[50];
-    char val[50];
-    bid128_to_string(val, x);
-    sprintf(buf, "%g", atof(val));
+    number_toString(x, buf, sizeof(buf));
     return buf;
 }
 
-/* ToDo: Implement proper number to string formatter. */
 void number_toString(Number x, char *buf, size_t len)
 {
-    //const int PRECISION = 34; /* ToDo: Finish implementation */
-    int iLeadingZeros = 0;
+    memset(buf, 0x00, len);
+    const int PRECISION = 34;
 
     assert(len != 0);
 
-    char val[50];
-    char lead[50];
+    char val[50] = { 0 };
     bid128_to_string(val, x);
 
-    char *s = &val[0];
-
-    /* Skip over the leading +, if there is one. */
-    while(*s != '\0' && (*s == '+')) s++;
-
-    char *v = s;
+    char *v, *s = v = &val[1];
 
-    while (*s != '\0' && (*s != 'E')) s++;
+    // Calculate the length of the string, locating the exponent value at the same time.
+    int slen = 0;
+    while (*s != '\0' && (*s != 'E')) s++, slen++;
 
     char *E = s;
 
-    if (*s == 'E') {
-        /* Deal with the exponent. */
-        int i;
-        iLeadingZeros = abs(atoi(++s));
-        for (i = 0; i < iLeadingZeros; i++) {
-            lead[i] = '0';
-        }
-        lead[i++] = '.';
-        lead[i++] = '\0';
-        strcpy(buf, lead);
+    if (*s != 'E') {
+        buf = val;
+        return;
     }
 
+    int exponent = atoi(++s);
     *E = '\0';
-    strcat(buf, v);
+    char *p = E - 1;
+
+    while ((p > v) && (*p == '0')) p--;
+
+    const int last_significant_digit = (int)(p - v) + 1;
+    const int trailing_zeros = (int)slen - last_significant_digit;
+    exponent += trailing_zeros;
+    char sbuf[50] = { 0 };
+    char r[50] = { 0 };
+    memcpy(sbuf, v, last_significant_digit);
+    slen = last_significant_digit;
+    sbuf[last_significant_digit+1] = '\0';
```

Comment:
Why is this setting `sbuf[last_significant_digit+1]`? There's a byte at `sbuf[last_significant_digit]` that wasn't set in the `memcpy`.

Reply:
Fixed.